app-marketplace-v2-spec/architecture.md

```diff
@@ -0,0 +1,146 @@
+      - [Widgets](#widgets)
+      - [Dialogs](#dialogs)
+    + [Change notifications.](#change-notifications)
+    + [Autocomplete lookups](#autocomplete-lookups)
+    + [Lifecycle - Install](#lifecycle---install)
+    + [Lifecycle Uninstall.](#lifecycle-uninstall)
+    + [Configuration.](#configuration)
+    + [mattermost-cloud-app.json](#mattermost-cloud-appjson)
+  * [mattermost-plugin-cloud-apps](#mattermost-plugin-cloud-apps)
+  * [mattermost-server](#mattermost-server)
+    + [Minimal Changes to the Core Code.](#minimal-changes-to-the-core-code)
+  * [Cloud App Marketplace](#cloud-app-marketplace)
+
+## Cloud Apps.
+
+### Authentication with Mattermost.
+1. OAuth2 will be used as a primary mechanism of authentication for the Mattermost REST API.
+2. Installing a **Cloud App** onto a Mattermost instance will provision sufficient credential to the app to invoke the REST API as the bot (priviledged? scoped? #TODO), and to verify/decode the JWT on the incoming requests from the Mattermost Server
+3. We should enable 3 use-cases:
+  a. Cloud App** using a "bot"-level token (to use APIs as the bot)
+  b. **Cloud App** using user-level OAuth2 tokens. Will require users' authoriation when first using the app.
+  c. "A trusted app" - all users will automatically get "on-behalf" tokens, like with Atlassian Connect. @crspeller mentioned a "trusted app" concept in OAuth2, need to research.
+4. Function requests, and Event notifications sent to the Cloud App will include a JWT. The JWT will contain the acting Mattermost UserID (or Bot UserID for non-user-specific notifications)
+
+### Authentication with Upstream Applications.
+1. Authentication with upstream applications will be left entirely to the Cloud App
+OR
+2. Mattermost will provide a "service" for mapping user accounts using OAuth2
+
+### Cloud App to Mattermost.
+Cloud Apps interact with Mattermost primarily by using the **Mattermost REST
+APIs**. The "Response" functionality of legacy command and action handlers will
+be greatly simplified for Cloud App functions. The cloud app functions will
+respond as either:
+  - 200 OK + (optional) simple markdown + (optional) JSON.
+  - Error code + error message.
+  - A (302?) "redirect" to navigate to a channel, external browser, an iframe
+    popup, or run another function.
+  - #TODO @larkox to reconcile with the "side effects" of his Integrations 2.0
+    spec.
+  - #TODO @levb, @larkox to propose redirect and execute response format.
+  - #TODO @crspeller, @lieut-data anything to borrow/learn from the workflow
+    projects.
+
+### Mattermost to Cloud App.
+Mattermost interacts with Cloud Apps in 3 ways: 
+  - **Functions** that are invoked via Interactive Inputs, and have
+    request/response semantics. They may be synchronous, or execute in the
+    background once invoked.
+  - 1-way **Change notifications** that are sent after changes occur (from the
+    server).
+  - **Autocomplete** queries.
+
+### Functions 
```

Comment:
Are functions registered by apps? What's the difference between function execution and calling an HTTP endpoint?

Reply:
The 3 types of HTTP points exposed by the app are ^^^ (functions, notifications, lookups).